feat(baselines) Add Floco Baseline (#4941)
adap/flower: baselines/floco/README.md

**Abstract:** Statistical heterogeneity in federated learning poses two major challenges: slow global training due to conflicting gradient signals, and the need of personalization for local distributions. In this work, we tackle both challenges by leveraging recent advances in *linear mode connectivity* --- identifying a linearly connected low-loss region in the parameter space of neural networks, which we call solution simplex. We propose federated learning over connected modes (Floco), where clients are assigned local subregions in this simplex based on their gradient signals, and together learn the shared global solution simplex. This allows personalization of the client models to fit their local distributions within the degrees of freedom in the solution simplex and homogenizes the update signals for the global simplex training.  Our experiments show that Floco accelerates the global training process, and significantly improves the local accuracy with minimal computational overhead in cross-silo federated learning settings.

## About this baseline

**What’s implemented:** The code in this directory implements the method Floco and Floco+.

**Datasets:** CIFAR-10

> **Hardware Setup:** We trained the models used for the plots below on a M1X chip.

Author: Change this to Apple M1 Pro chip.

---
title: Federated Learning over Connected Modes
url: https://openreview.net/pdf?id=JL2eMCfDW8

> labels: [personalization] 

Reviewer: @dennis-grinwald, what would be two other labels to use for `Floco`? I'm thinking of `cross-silo` and `simplex`

Author: I'd say let's use cross-silo and linear mode connectivity :)

Author: And thank you for putting in the work! Looking forward to seeing the method merged :)
If there is anything I could do to help, let me know!

Reviewer: perfect, i added those labels

> labels: [personalization] 

Reviewer: ```suggestion
labels: [personalization, cross-silo, linear mode connectivity] 
```

> dataset: [CIFAR10]

Reviewer: ```suggestion
dataset: [CIFAR-10]
```

## About this baseline

**What’s implemented:** The code in this directory implements the method Floco and Floco+.

**Datasets:** CIFAR-10

**Hardware Setup:** We trained the models used for the plots below on an Apple M1 Pro chip.

> **Contributors:** <add here your names (or github tags, emails are fine too if you'd like)> 

Reviewer: I think this is the last remaining bit to add to the README.md.

Author: Changed this.

## Experimental Setup

**Task:** Image Classification

> **Model:** This repository implements a two-layer CNN as in \cite{FedAvg}. Additional models can be implemented by changing the `base_model` attribute in the `SimplexModel` class in `floco/model.py`.

Reviewer: ```suggestion
**Model:** This repository implements a two-layer CNN as in [McMahan et al. 2017](https://arxiv.org/abs/1602.05629). Additional models can be implemented by changing the `base_model` attribute in the `SimplexModel` class in `floco/model.py`.
```

Reviewer: this is what you meant by `\cite{FedAvg}` right?

Author: Yes, correct. Thanks!
